packages/core/src/model-manager.ts

```diff
@@ -1,40 +1,49 @@
 'use strict';
 
-const Toposort = require('toposort-class');
-const _ = require('lodash');
+import defaults from 'lodash/defaults';
+import Toposort from 'toposort-class';
+import type { ModelStatic } from './model';
+import type { Sequelize } from './sequelize';
 
 export class ModelManager {
-  constructor(sequelize) {
+  declare sequelize: Sequelize;
+  declare models: ModelStatic[];
+
+  constructor(sequelize: Sequelize) {
     this.models = [];
     this.sequelize = sequelize;
   }
 
-  addModel(model) {
+  addModel<T extends ModelStatic>(model: T): T {
     this.models.push(model);
     this.sequelize.models[model.name] = model;
 
     return model;
   }
 
-  removeModel(modelToRemove) {
-    this.models = this.models.filter(model => model.name !== modelToRemove.name);
+  removeModel(modelToRemove: ModelStatic): void {
+    this.models = this.models.filter(
+      model => model.name !== modelToRemove.name,
+    );
 
     delete this.sequelize.models[modelToRemove.name];
   }
 
-  getModel(modelName) {
+  getModel(modelName: string): ModelStatic | undefined {
     return this.models.find(model => model.name === modelName);
   }
 
-  findModel(callback) {
+  findModel(
+    callback: (model: ModelStatic) => boolean,
+  ): ModelStatic | undefined {
```

Comment:
Would it be better to be able to pass a type for the model static to the function to return a specific rather than generic model static instance?

```suggestion
  findModel<T extends ModelStatic>(
    callback: (model: ModelStatic) => boolean,
  ): T | undefined {
```

Reply:
Agreed.

Reply:
Q: I submitted the suggestion but now it fails with the following

![image](https://github.com/sequelize/sequelize/assets/79163/b6e757a5-6e9f-4517-a75c-709d2fc905ef)

Why do we prefer generic here and why does it not matter for getModel?

Also, does this look like a decent workaround?

```
    return this.models.find(callback) as T | undefined;

```

Reply:
hmm actually in this instance there should be no generic at all

Reply:
So I revert the suggestion and go with ModelStatic instead?

Reply:
Yes, the actual type if we want to filter the type based on the callback is much more complicated and not worth it for the rare use cases this has

Reply:
reverted